test/e2e/common.go

```diff
@@ -215,3 +217,39 @@ func createRestConfig(ctx context.Context, tmpdir, namespace, clusterName string
 
 	return config
 }
+
+// EnsureControlPlaneInitialized waits for the cluster KubeadmControlPlane object to be initialized
+// and then installs cloud-provider-azure components via Helm.
+// Fulfills the clusterctl.Waiter type so that it can be used as ApplyClusterTemplateAndWaitInput data
+// in the flow of a clusterctl.ApplyClusterTemplateAndWait E2E test scenario.
+func EnsureControlPlaneInitialized(ctx context.Context, input clusterctl.ApplyClusterTemplateAndWaitInput, result *clusterctl.ApplyClusterTemplateAndWaitResult) {
+	getter := input.ClusterProxy.GetClient()
+	cluster := framework.GetClusterByName(ctx, framework.GetClusterByNameInput{
+		Getter:    getter,
+		Name:      input.ConfigCluster.ClusterName,
+		Namespace: input.ConfigCluster.Namespace,
+	})
+	kubeadmControlPlane := &kubeadmv1.KubeadmControlPlane{}
+	key := crclient.ObjectKey{
+		Namespace: cluster.Spec.ControlPlaneRef.Namespace,
+		Name:      cluster.Spec.ControlPlaneRef.Name,
+	}
+	Eventually(func() error {
+		return getter.Get(ctx, key, kubeadmControlPlane)
+	}, input.WaitForControlPlaneIntervals...).Should(Succeed(), "Failed to get KubeadmControlPlane object %s/%s", cluster.Spec.ControlPlaneRef.Namespace, cluster.Spec.ControlPlaneRef.Name)
+	if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["cloud-provider"] == "external" {
+		InstallCloudProviderAzureHelmChart(ctx, input)
+	}
+	if kubeadmControlPlane.Spec.KubeadmConfigSpec.ClusterConfiguration.ControllerManager.ExtraArgs["external-cloud-volume-plugin"] == "azure" {
```

Comment:
@sonasingh46 to approve this being a definitive configuration trigger for installing the azuredisk CSI helm chart

Reply:
This flag is necessary during upgrade to kubernetes version 1.23 that uses external cloud provider otherwise the upgrade will fail. An upgrade to k8s version 1.23 will mean to install external azurediskcsi driver as the azurecsidisk migration is enabled by default. 

So wether external azurediskcsi driver should be installed or not should solely depend on if AzureDiskCSIMigraion is enabled or not -- 
i) For k8s version 1.23 and later it is enabled by default.
ii) For k8s version prior to 1.23, it can be enabled by passing a flag.

TL;DR
At this place, I think it is more helpful if we leave upon the users/callers to decide what helm component they want to install by becoming more explicit. For example, let us say we have 5 helm things -- but as per the specific CI template only 3  installs is required.

Reply:
One simple way top of my head is something like following: 

```
const (
	AzureDiskCSIDriver = "AzureDiskCSIDriver"
	CloudProviderAzure = "CloudProviderAzure"
)

var Installers = map[string]func(){
	AzureDiskCSIDriver : InstallAzureDiskCSIDriver(),
	CloudProviderAzure : InstallCloudProviderAzure(),
}

func InstallAzureDiskCSIDriver() func(){
	return func() {
		// install code
	}
}

func InstallCloudProviderAzure() func(){
	return func() {
		// install code
	}
}

type HelmInstallOpts struct{
	Packages []string
}

func EnsureControlPlaneInitialized(opts HelmInstallOpts){
	// Already existing code
	for _,p:= range opts.Packages{
		install,ok:=Installers[p]
		if !ok{
			// report error as p is unsupported helm package
		}
		install()
	}
}

```
wdyt? Or any approach that just gives the flexibility to choose what helm to install.

Reply:
We can do this in a different PR also.

Reply:
I love that you're thinking of users here, but in this case I think we are focused solely on figuring out how to evaluate from a given cluster template yaml "am I able to install (or do I need to install) this particular addon?" for CI testing only.

I seems to me like the azuredisk-csi drivers actually work in any Kubernetes cluster scenario for the versions that we test, so how about we just install it for every CI test template?

Reply:
Okay. So I am coming from a point that -- there are cases where we do not want to install azurediskcsi driver for e2e. 
Like an e2e for workload cluster upgrade from v1.22 to v1.23.

Reply:
But the flag check should do fine. Let us go with it and if need arises we will revisit.